File: signalling/clients/drop.py
#!/usr/bin/env python3
"""Weizen P2P drop — Python (aiortc) client for the token-less signalling worker.

ต่อ wss://weizen-p2p-signalling.weizen.workers.dev/room/<room> (open relay, ไม่มี token)
แล้วคุย WebRTC DataChannel ตรงระหว่าง peer — ส่งไฟล์เป็น binary chunks (E2E).
aiortc ใส่ ICE candidate ใน SDP ให้เลย (non-trickle) → แลกแค่ offer/answer ผ่าน worker พอ.

ติดตั้ง:  python3 -m venv .venv && . .venv/bin/activate && pip install aiortc websockets
ใช้:
  python3 drop.py recv --room arra --name weizen-recv
  python3 drop.py send --room arra --name weizen --to weizen-recv --file ./hello.txt
"""
import asyncio, json, os, argparse
import websockets
from aiortc import RTCPeerConnection, RTCSessionDescription

SIGNAL = "wss://weizen-p2p-signalling.weizen.workers.dev/room/"
ICE_CFG = None  # aiortc ใช้ STUN default; ตั้ง RTCConfiguration ได้ถ้าต้องการ TURN
CHUNK = 16 * 1024


async def run(mode, room, name, to=None, path=None):
    ws = await websockets.connect(SIGNAL + room, max_size=None)
    pcs = {}              # peerId -> RTCPeerConnection
    self_id = None
    done = asyncio.Event()

    async def send(obj): await ws.send(json.dumps(obj))

    def new_pc(peer):
        pc = RTCPeerConnection()
        pcs[peer] = pc
        return pc

    async def wire_recv(channel):
        meta = {}
        f = None
        @channel.on("message")
        def on_msg(msg):
            nonlocal f, meta
            if isinstance(msg, str):
                m = json.loads(msg)
                if m.get("kind") == "file-meta":
                    meta = m
                    os.makedirs("inbox", exist_ok=True)
                    f = open(os.path.join("inbox", os.path.basename(m["name"])), "wb")
                    print(f"[recv] start {m['name']} ({m['size']} B)")
                elif m.get("kind") == "file-end":
                    if f: f.close()
                    print(f"[recv] ✅ saved inbox/{os.path.basename(meta.get('name','file'))}")
                    done.set()
            else:
                if f: f.write(msg)

    async def do_send(peer):
        pc = new_pc(peer)
        ch = pc.createDataChannel("weizen")
        @ch.on("open")
        async def on_open():
            data = open(path, "rb").read()
            ch.send(json.dumps({"kind": "file-meta", "name": os.path.basename(path), "size": len(data)}))
            for i in range(0, len(data), CHUNK):
                ch.send(data[i:i + CHUNK])
            ch.send(json.dumps({"kind": "file-end", "name": name}))
            print(f"[send] ✅ sent {os.path.basename(path)} ({len(data)} B) -> {peer}")
            await asyncio.sleep(0.5)
            done.set()
        offer = await pc.createOffer()
        await pc.setLocalDescription(offer)
        await send({"type": "offer", "to": peer, "sdp": {"type": pc.localDescription.type, "sdp": pc.localDescription.sdp}})

    async for raw in ws:
        m = json.loads(raw)
        t = m.get("type")
        if t == "welcome":
            self_id = m["self"]
            print(f"[ok] connected as {name} ({self_id}); peers={m['peers']}")
            if mode == "send":
                # หา target จาก peers ที่ออนไลน์ (ในเดโมนี้ join ก่อน = เห็นใน peers)
                if not m["peers"]:
                    print("[send] ยังไม่มี peer ในห้อง — ให้ผู้รับ join ก่อน"); break
                await do_send(m["peers"][-1])
        elif t == "peer-join" and mode == "send" and path:
            await do_send(m["from"])
        elif t == "offer":
            pc = new_pc(m["from"])
            @pc.on("datachannel")
            def on_dc(channel): asyncio.ensure_future(wire_recv(channel))
            await pc.setRemoteDescription(RTCSessionDescription(**m["sdp"]))
            ans = await pc.createAnswer()
            await pc.setLocalDescription(ans)
            await send({"type": "answer", "to": m["from"], "sdp": {"type": pc.localDescription.type, "sdp": pc.localDescription.sdp}})
        elif t == "answer":
            await pcs[m["from"]].setRemoteDescription(RTCSessionDescription(**m["sdp"]))
        if done.is_set():
            break

    for pc in pcs.values(): await pc.close()
    await ws.close()


if __name__ == "__main__":
    ap = argparse.ArgumentParser()
    ap.add_argument("mode", choices=["send", "recv"])
    ap.add_argument("--room", required=True)
    ap.add_argument("--name", required=True)
    ap.add_argument("--to")
    ap.add_argument("--file")
    a = ap.parse_args()
    asyncio.run(run(a.mode, a.room, a.name, a.to, a.file))

File: signalling/clients/drop.ts
#!/usr/bin/env bun
/**
 * Weizen P2P drop — Bun + werift CLI (token-less).
 * ต่อ open signalling worker → WebRTC DataChannel → ส่งไฟล์ตรงระหว่าง peer (E2E).
 * werift = pure-JS WebRTC (ไม่ต้อง native build) · non-trickle (ICE ฝังใน SDP).
 *
 * ติดตั้ง:  bun install      (ลง werift)
 * ใช้:
 *   bun drop.ts recv --room arra --name weizen-recv
 *   bun drop.ts send --room arra --name weizen --to weizen-recv --file ./hello.txt
 *   (--to ละได้ = auto-pick peer แรกในห้อง · SIGNAL_URL override ได้ผ่าน env)
 */
import { RTCPeerConnection } from "werift";
import { writeFileSync, readFileSync, mkdirSync } from "fs";
import { basename, join } from "path";

const SIGNAL = process.env.SIGNAL_URL || "wss://weizen-p2p-signalling.weizen.workers.dev/room/";
const STUN = { iceServers: [{ urls: "stun:stun.l.google.com:19302" }, { urls: "stun:stun.cloudflare.com:3478" }] };
const CHUNK = 16 * 1024;

const mode = process.argv[2];
const arg = (k: string) => { const i = process.argv.indexOf("--" + k); return i >= 0 ? process.argv[i + 1] : undefined; };
const room = arg("room") || "weizen-lobby";
const name = arg("name") || "weizen-" + Math.floor(Math.random() * 9999);
const to = arg("to");
const file = arg("file");

if (mode !== "recv" && mode !== "send") {
  console.log("usage: bun drop.ts <recv|send> --room <r> --name <n> [--to <peer> --file <path>]");
  process.exit(1);
}

async function gather(pc: any, desc: any) {
  await pc.setLocalDescription(desc);
  if (pc.iceGatheringState !== "complete") {
    await new Promise<void>((res) => pc.iceGatheringStateChange.subscribe((s: string) => s === "complete" && res()));
  }
  return pc.localDescription;
}

const ws = new WebSocket(SIGNAL + encodeURIComponent(room));
const sendWs = (o: any) => ws.send(JSON.stringify(o));
const pcs = new Map<string, any>();

function recvChannel(dc: any) {
  let meta: any = {}; const buf: Buffer[] = [];
  dc.onMessage.subscribe((data: any) => {
    if (typeof data === "string") {
      const m = JSON.parse(data);
      if (m.kind === "file-meta") { meta = m; console.log(`[recv] start ${m.name} (${m.size} B)`); }
      else if (m.kind === "file-end") {
        mkdirSync("inbox", { recursive: true });
        const out = join("inbox", basename(meta.name || "file"));
        writeFileSync(out, Buffer.concat(buf));
        console.log(`[recv] ✅ saved ${out} (${Buffer.concat(buf).length} B)`);
      }
    } else { buf.push(Buffer.from(data)); }
  });
}

async function sendTo(peer: string) {
  const pc = new RTCPeerConnection(STUN as any); pcs.set(peer, pc);
  const dc = pc.createDataChannel("weizen");
  dc.stateChanged.subscribe((st: string) => {
    if (st === "open") {
      const data = readFileSync(file!);
      dc.send(JSON.stringify({ kind: "file-meta", name: basename(file!), size: data.length }));
      for (let i = 0; i < data.length; i += CHUNK) dc.send(data.subarray(i, i + CHUNK));
      dc.send(JSON.stringify({ kind: "file-end", name }));
      console.log(`[send] ✅ sent ${basename(file!)} (${data.length} B) → ${peer}`);
      setTimeout(() => process.exit(0), 800);
    }
  });
  const offer = await gather(pc, await pc.createOffer());
  sendWs({ type: "offer", to: peer, sdp: { type: offer.type, sdp: offer.sdp } });
}

ws.onopen = () => console.log(`[${mode}] connecting room "${room}" as ${name}…`);
ws.onmessage = async (ev: any) => {
  const m = JSON.parse(ev.data);
  if (m.type === "welcome") {
    console.log(`[ok] ${name} (${m.self}) · peers=${JSON.stringify(m.peers)}`);
    if (mode === "send") {
      const target = to && m.peers.includes(to) ? to : m.peers[m.peers.length - 1];
      if (!target) { console.log("[send] ยังไม่มี peer ในห้อง — ให้ผู้รับ join ก่อน"); process.exit(1); }
      await sendTo(target);
    }
  } else if (m.type === "peer-join" && mode === "send" && file && to === m.from) {
    await sendTo(m.from);
  } else if (m.type === "offer") {
    const pc = new RTCPeerConnection(STUN as any); pcs.set(m.from, pc);
    pc.onDataChannel.subscribe((dc: any) => recvChannel(dc));
    await pc.setRemoteDescription(m.sdp);
    const answer = await gather(pc, await pc.createAnswer());
    sendWs({ type: "answer", to: m.from, sdp: { type: answer.type, sdp: answer.sdp } });
  } else if (m.type === "answer") {
    await pcs.get(m.from)?.setRemoteDescription(m.sdp);
  }
};
if (mode === "recv") console.log("[recv] เฝ้ารับไฟล์ลง ./inbox — Ctrl-C เพื่อออก");
